src/mono/mono/mini/debugger-agent.c

```diff
@@ -7504,14 +7524,25 @@ assembly_commands (int command, guint8 *p, guint8 *end, Buffer *buf)
 		break;
 	}
 	case MDBGPROT_CMD_ASSEMBLY_GET_PEIMAGE_ADDRESS: {
-        MonoImage* image = ass->image;
-        if (ass->dynamic) {
-            return ERR_NOT_IMPLEMENTED;
-        }
-		buffer_add_long (buf, (guint64)(gsize)image->raw_data);
-        buffer_add_int (buf, image->raw_data_len);
+		MonoImage* image = ass->image;
+		if (ass->dynamic) {
+		    return ERR_NOT_IMPLEMENTED;
+		}
+		// Mdbg uses arithmetics with this pointer and RVA to get information using readmemory, 
+		// but it doesn't work on mono, it should call mono_cli_rva_image_map to get the right offset and don't use pure RVA.
+		// To workaround it, I'm creating a buffer with the full file content and returning this address, this is not okay for memory
+		// usage, but it will work for now.
+		HMODULE module_handle;
+		gunichar2 *file16;
+		file16 = u8to16(mono_image_get_filename (ass->image)); 
+		module_handle = LoadLibrary (file16);
+		g_free(file16);
+
+		PRINT_DEBUG_MSG(1, "MDBGPROT_CMD_ASSEMBLY_GET_PEIMAGE_ADDRESS - [%p] - %d\n", module_handle, image->raw_data_len);
+		buffer_add_long (buf, (gssize)module_handle);
+		buffer_add_int (buf, image->raw_data_len);
```

Comment:
Discussed offline.  We should try to find another way instead of using this hack.  Maybe mscordbi can do address translation instead.

(The reason `LoadLibrary` works and just `image->raw_data` doesn't, is because when you load a PE image you're supposed to write some bytes to one of its headers (I am not a PE expert, so I'm just guessing here). So when mdbg pokes aroudn in memory, it will do the RVA translation based on those bytes.  But Mono doesn't do any writing when opening MonoImages - we try to mmap them without writing anything.  `LoadLibrary` on the other hand does write the right bytes.)

Reply:
I changed on mdbg side, let's see if they approve the change.

Reply:
@thaystg I was thinking - if we know what data in the PE header is changed when it is loaded by `LoadLibrary`, we could store a modified copy of the header somewhere that the debugger can use.  Then when we get a "read memory" request in the debugger - if they're reading in the range of the PE header, we look in the modified version.  (I would like to avoid loading most of the MonoImage read/write, but if it's only a couple of pages that are just needed for the debugger, we could memcpy and modify those)

Reply:
I'm not sure about it, when I read with LoadLibrary and in our raw_data, the offset that we want to read is 8096 in both cases, so I don't think that LoadLibrary changes the header, what I think is that loading with LoadLibrary the offset used in memory is exactly the same that is written in header. 
Of course that we could try to fix the header with our new offset, this is a valid solution also.
I don't think that the debugger uses more than the header, and the header is okay, that's why I didn't try to find an excellent solution for it. When I did the tests using VsCode and Vs they just need to read the header, and this was okay.

Reply:
>  so I don't think that LoadLibrary changes the header, what I think is that loading with LoadLibrary the offset used in memory is exactly the same that is written in header.

Ah ok.  That makes a lot of sense.

> Of course that we could try to fix the header with our correspondent offset, this is a valid solution also.

Let's see if they'll take changes to mdbg.  If they do not like it, we can keep this as one of the alternatives that we can try.

Reply:
We figured out that mdbg has code to support PE images either in disk or memory layout (the difference being in how RVAs are resolved to absolute addresses - memory layout it's a simple addition, but disk layout has to respect coff sections).  But they were using a heuristic that was only valid for CoreCLR.  But Thays found that `ICorDebugModule4::IsMappedLayout` is a better guide for which layout to use (mapped == disk, flat == memory.  mono always uses mapped; coreclr mostly uses flat, except for Load(byte[]) assemblies).

https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/coreclr/inc/cordebug.idl#L5203